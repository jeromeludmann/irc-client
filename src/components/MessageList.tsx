import React from "react";

interface Props {
  messages: string[];
}

export default ({ messages }: Props) => (
  // TODO do not use index
  <ul
    style={{
      fontFamily: "Menlo, Monaco, Courier",
      fontSize: "12px",
      listStyle: "none",
      padding: "0",
    }}
  >
    {messages.map((message, i) => <li key={i}>{message}</li>)}
  </ul>
);
