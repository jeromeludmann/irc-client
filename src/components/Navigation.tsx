import React, { PureComponent, MouseEvent, CSSProperties } from "react";

interface Props {
  servers: {
    [s: string]: {
      windows: { [w: string]: { unread: boolean } };
    };
  };
  active: {
    server: string;
    window: string;
  };
  onWindowButtonClick: (
    { server, window }: { server: string; window: string },
  ) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { active, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(server => (
          <ul key={server} style={{ listStyle: "none", padding: "0" }}>
            {Object.keys(servers[server].windows).map(window => {
              const isActive =
                server === active.server && window === active.window;
              const isUnread = servers[server].windows[window].unread;
              const style: CSSProperties = {
                outline: "none",
                border: "1px solid #ddd",
                width: "100px",
                color: isActive ? "black" : "#bbb",
                backgroundColor: isUnread ? "yellow" : "#fff",
              };

              return (
                <li key={window} style={{ marginBottom: "5px" }}>
                  <button
                    onClick={this.handleClick}
                    data-server={server}
                    data-window={window}
                    style={style}
                  >
                    {window}
                  </button>
                </li>
              );
            })}
          </ul>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const server = e.currentTarget.getAttribute("data-server");
    const window = e.currentTarget.getAttribute("data-window");

    if (server && window) {
      this.props.onWindowButtonClick({ server, window });
    }
  };
}
