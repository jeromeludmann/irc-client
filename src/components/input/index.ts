import { MapStateToProps, connect } from 'react-redux'
import { RootState } from '@app/reducers'
import Input from '@app/components/input/Input'
import {
  updateInputValue,
  enterInputValue,
  goBackInputHistory,
  goForwardInputHistory,
} from '@app/actions/ui'
import { selectInputValue } from '@app/reducers/selectors'

interface StateProps {
  value: string
}

interface DispatchProps {
  onType: (value: string) => void
  onEnter: (value: string) => void
  onArrowUp: () => void
  onArrowDown: () => void
}

const mapStateToProps: MapStateToProps<StateProps, {}, RootState> = state => ({
  value: selectInputValue(state),
})

const mapDispatchToProps: DispatchProps = {
  onType: updateInputValue,
  onEnter: enterInputValue,
  onArrowUp: goBackInputHistory,
  onArrowDown: goForwardInputHistory,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Input)
