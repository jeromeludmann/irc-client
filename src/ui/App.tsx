import React from "react";
import ConnectedServerList from "@app/ui/tmp/ConnectedServerList";

export default class App extends React.Component {
  public render() {
    return <ConnectedServerList />; // for test only
  }
}
