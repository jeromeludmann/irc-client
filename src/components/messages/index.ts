import { connect, MapStateToProps } from "react-redux";
import MessageList from "@app/components/messages/MessageList";
import { MessagesState } from "@app/reducers/messages";
import { RootState } from "@app/reducers";
import { selectMessages } from "@app/reducers/selectors";

interface StateProps {
  messages: MessagesState;
}

const mapStateToProps: MapStateToProps<StateProps, {}, RootState> = state => ({
  messages: selectMessages(state),
});

export default connect(mapStateToProps)(MessageList);
