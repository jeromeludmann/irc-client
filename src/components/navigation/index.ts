import { connect, MapStateToProps } from 'react-redux'
import { RouteState } from '@app/reducers/route'
import { Route } from '@app/utils/Route'
import { RootState } from '@app/reducers'
import { switchWindow } from '@app/actions/ui'
import Navigation from '@app/components/navigation/Navigation'
import { selectServers, selectRoute } from '@app/reducers/selectors'

interface StateProps {
  servers: {
    [serverKey: string]: {
      name: string
      buffers: { [bufferKey: string]: { activity: boolean } }
    }
  }
  window: RouteState
}

interface DispatchProps {
  onWindowButtonClick: (route: Route) => void
}

const mapStateToProps: MapStateToProps<StateProps, {}, RootState> = state => ({
  // TODO provide a better selector to Navigation
  servers: selectServers(state),
  window: selectRoute(state),
})

const mapDispatchToProps: DispatchProps = {
  onWindowButtonClick: switchWindow,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Navigation)
