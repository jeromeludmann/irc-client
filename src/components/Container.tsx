import React, { Component } from "react";
import { connect, MapStateToProps } from "react-redux";
import { setWindow } from "@app/actions/ui/active-window";
import { RootState } from "@app/reducers";
import { ServerRouterState } from "@app/reducers/server-router";
import { ActiveWindowState } from "@app/reducers/active";
import Navigation from "@app/components/Navigation";
import MessageList from "@app/components/MessageList";
import Input from "@app/components/Input";
import {
  selectActiveWindow,
  selectServers,
  selectUser,
} from "@app/reducers/selectors";
import { MessageListState } from "@app/reducers/window/messages";
import { selectMessages } from "@app/reducers/window/selectors";
import { selectValue } from "@app/reducers/input/selectors";
import {
  changeInputValue,
  sendInputValue,
  goBackInputHistory,
  goForwardInputHistory,
} from "@app/actions/ui/input";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";

interface StateProps {
  user: UserState;
  servers: ServerRouterState;
  active: ActiveWindowState;
  messages: MessageListState;
  inputValue: string;
}

interface DispatchProps {
  onWindowSwitch: (route: Route) => void;
  onInputType: (value: string) => void;
  onInputEnter: (value: string) => void;
  onInputArrowUp: () => void;
  onInputArrowDown: () => void;
}

class Container extends Component<StateProps & DispatchProps> {
  public render() {
    return (
      <>
        <div>nick: {this.props.user.nick}</div>

        <Navigation
          servers={this.props.servers}
          active={this.props.active}
          onWindowButtonClick={this.props.onWindowSwitch}
        />

        <Input
          value={this.props.inputValue}
          onType={this.props.onInputType}
          onEnter={this.props.onInputEnter}
          onArrowUp={this.props.onInputArrowUp}
          onArrowDown={this.props.onInputArrowDown}
        />

        <MessageList messages={this.props.messages} />
      </>
    );
  }
}

const mapStateToProps: MapStateToProps<StateProps, {}, RootState> = state => ({
  user: selectUser(state),
  messages: selectMessages(state),
  active: selectActiveWindow(state),
  servers: selectServers(state),
  inputValue: selectValue(state),
});

const mapDispatchToProps: DispatchProps = {
  onWindowSwitch: setWindow,
  onInputType: changeInputValue,
  onInputEnter: sendInputValue,
  onInputArrowUp: goBackInputHistory,
  onInputArrowDown: goForwardInputHistory,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Container);
