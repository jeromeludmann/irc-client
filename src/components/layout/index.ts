import { connect, MapStateToProps } from "react-redux";
import { AppState } from "@app/reducers";
import { ServersState } from "@app/reducers/servers";
import { RouteState } from "@app/reducers/route";
import { selectRoute, selectServers } from "@app/reducers/_selectors";
import { MessagesState } from "@app/reducers/channel/messages";
import { selectMessages } from "@app/reducers/channel/_selectors";
import { selectValue } from "@app/reducers/input/_selectors";
import {
  updateInputValue,
  enterInputValue,
  goBackInputHistory,
  goForwardInputHistory,
  switchWindow,
} from "@app/actions/ui";
import { Route } from "@app/Route";
import { Layout } from "@app/components/layout/Layout";

interface StateProps {
  servers: ServersState;
  window: RouteState;
  messages: MessagesState;
  inputValue: string;
}

interface DispatchProps {
  onChannelSwitch: (route: Route) => void;
  onInputType: (value: string) => void;
  onInputEnter: (value: string) => void;
  onInputArrowUp: () => void;
  onInputArrowDown: () => void;
}

const mapStateToProps: MapStateToProps<StateProps, {}, AppState> = state => ({
  messages: selectMessages(state),
  window: selectRoute(state),
  servers: selectServers(state),
  inputValue: selectValue(state),
});

const mapDispatchToProps: DispatchProps = {
  onChannelSwitch: switchWindow,
  onInputType: updateInputValue,
  onInputEnter: enterInputValue,
  onInputArrowUp: goBackInputHistory,
  onInputArrowDown: goForwardInputHistory,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Layout);
