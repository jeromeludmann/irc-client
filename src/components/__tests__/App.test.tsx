import React from "react";
import { shallow } from "enzyme";
import { App } from "@app/components/App";

describe("App component", () => {
  it("should render correctly", () => {
    expect(shallow(<App />)).toMatchSnapshot();
  });
});
