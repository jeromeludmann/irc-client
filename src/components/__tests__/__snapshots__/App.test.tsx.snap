// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <styled.nav>
          <Connect(Navigation) />
        </styled.nav>,
        <styled.div>
          <styled.div>
            <Connect(Component) />
          </styled.div>
          <styled.div>
            <Connect(Input) />
          </styled.div>
        </styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Connect(Navigation) />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.div>
              <Connect(Component) />
            </styled.div>,
            <styled.div>
              <Connect(Input) />
            </styled.div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Connect(Component) />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Connect(Input) />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <styled.nav>
            <Connect(Navigation) />
          </styled.nav>,
          <styled.div>
            <styled.div>
              <Connect(Component) />
            </styled.div>
            <styled.div>
              <Connect(Input) />
            </styled.div>
          </styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Connect(Navigation) />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.div>
                <Connect(Component) />
              </styled.div>,
              <styled.div>
                <Connect(Input) />
              </styled.div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Connect(Component) />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Connect(Input) />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
